cassiopeia/simulator/TumorWithAFitSubclone.py

```diff
@@ -0,0 +1,141 @@
+    phylogenies and (2) the phylogenies are as simple as possible while being
+    non-neutral. The ability to control the branch lengths of the neutrally
+    evolving cells and of the fit cells allows us to simulate different degrees
+    of fitness.
+
+    Note that the branches for the leaves of the tree need not have length
+    exactly equal to branch_length_neutral nor branch_length_fit, because the
+    cells get assayed at the arbitrary time specified by 'experiment_duration',
+    which is not necessarily the time at which they divide.
+
+    The names of the leaf nodes will be given by a unique integer ID, followed
+    by "_neutral" for the neutrally evolving cells, and by "_fit" for the fit
+    cells.
+
+    Finally, note that we do not enforce the condition branch_length_neutral >
+    branch_length_fit, therefore it is possible to make the subclone less fit,
+    i.e. expand slower.
+
+    Args:
+        branch_length_neutral: Branch length of the neutrally evolving cells.
+            All cells are neutrally evolving until generation
+            'generations_until_fit_subclone', when exactly one of the lineages
+            develops fitness.
+        branch_length_fit: The branch length of the fit subclone, which appears
+            exactly at generation 'generations_until_fit_subclone'.
+        experiment_duration: The total length of the experiment.
+        generations_until_fit_subclone: The generation at which one lineage
+            develops fitness, giving rise to a fit subclone.
+    """
+
+    def __init__(
+        self,
+        branch_length_neutral: float,
+        branch_length_fit: float,
+        experiment_duration: float,
+        generations_until_fit_subclone: int,
+    ):
+        self.branch_length_neutral = branch_length_neutral
+        self.branch_length_fit = branch_length_fit
+        self.experiment_duration = experiment_duration
+        self.generations_until_fit_subclone = generations_until_fit_subclone
+
+    def simulate_tree(self) -> CassiopeiaTree:
+        r"""
+        See base class.
+        """
+        branch_length_neutral = self.branch_length_neutral
+        branch_length_fit = self.branch_length_fit
+        experiment_duration = self.experiment_duration
+        generations_until_fit_subclone = self.generations_until_fit_subclone
+
+        def node_name_generator():
+            i = 0
+            while True:
+                yield str(i)
+                i += 1
+
+        tree = nx.DiGraph()  # This is what will get populated.
+
+        names = node_name_generator()
+        q = Queue()  # Contains: (node, time, fitness, generation)
+        times = {}
+
+        root = next(names) + "_neutral"
+        tree.add_node(root)
+        times[root] = 0.0
+
+        root_child = next(names) + "_neutral"
+        tree.add_edge(root, root_child)
+        q.put((root_child, 0.0, "neutral", 0))
+        subclone_started = False
```

Comment:
Why do you add a single child here? I would have imagined adding two children from the root. Why not just add the root directly to the queue?

Reply:
Because the founding cell doesn't divide immediately.

Reply:
(In particular the resulting tree is not binary)

Reply:
Sorry if I misunderstood the question BTW, maybe we should chat about it.

Reply:
You're totally right - the founding cell doesn't divide immediately, but it also doesn't affect the topology at all.

The time that it takes for a cell to divide can be encoded in the branch lengths leading to both of its children, right? In general, I think the simulations should return a tree that is well-formed, in the sense that each node should have at least 2 children (unless it's a leaf). Does that sound reasonable?

Reply:
I love these subtleties here! Let me try to clarify what's on my mind. My interpretation of a single-cell phylogeny is as follows:
- The root of the tree is the "founding cell" event.
- An internal node of the phylogeny which is not root, corresponds to a cell division event.
- Each branch of the phylogeny corresponds to the lifetime of a cell.
With this interpretation:
- The branch lengths leading from the root to each of its children are the times it takes for each of the daughter cells to divide - which is distinct from the time it takes for the founding cell to divide.
- The time it takes for the founding cell to divide is encoded in the first branch of the tree.

Regarding the comment that `each node should have at least 2 children`, I agree with this for internal nodes different from the root, but the way I interpret the phylogeny, the root should have degree 1, and in my mind this reflects the biological process more accurately. Here's one question to put my interpretation into perspective: where on the phylogeny should the CRISPR/Cas9 mutations that hit the founding cell - before it divides - be mapped to, if the root is required to have two children?

I'm happy to talk about this in more depth!

Reply:
For posterity: we decided that having a founder cell with a single child makes more sense from a modeling perspective. It doesn't change the tree topology, but will have several downstream ramifications when overlaying the lineage tracing data and inferring branch lengths. For this reason we're keeping it here.